Cloud9 Support for Resource-Importing Lab (#223)
aws-samples/cfn101-workshop: content/intermediate/operations/resource-importing/index.md

                          "ParameterKey": "Topic2Name",
                          "ParameterValue": "Topic2"
                      },
                      {
                          "ParameterKey": "Topic1Name",
                          "ParameterValue": "Topic1"
                      }
                  ],
                  "CreationTime": "2023-05-17T01:00:50.284000+00:00",
                  "LastUpdatedTime": "2023-05-17T01:35:38.408000+00:00",
                  "RollbackConfiguration": {},
                  "StackStatus": "IMPORT_COMPLETE",
                  "DisableRollback": false,
                  "NotificationARNs": [],
                  "Tags": [],
                  "EnableTerminationProtection": false,
                  "DriftInformation": {
                      "StackDriftStatus": "NOT_CHECKED"
                  }
              }
          ]
      }
      :::
    ::::
	::::tab{id="local" label="Local development"}  

    1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
    1. Select the stack named `cfn-workshop-resource-importing` and, from **Stack actions**, choose **Import resources into stack**.

>     1. Read the **Import Overview** and choose **Next**.

Reviewer: ```suggestion
    1. Read the **What you'll need** information and choose **Next**.
```

    1. From **Specify template**, choose **Upload a template file**. Upload the `resource-importing.yaml` template you updated with this lab part, and choose **Next**.

>     1. For the [**Identifier Value**](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import.html#resource-import-overview), specify the topic ARN value you noted after you created `Topic2`.

Reviewer: ```suggestion
    1. For the [**Identifier Value**](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import.html#resource-import-overview), specify the topic ARN value you noted after you created `Topic2`, and choose **Next**.
```

        --template-body file://resource-importing.yaml \
        --parameters ParameterKey=Topic2Name,ParameterValue=Topic2
      :::
      1. Review the change set to make sure the correct resources are removed.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation describe-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Execute the change set to import the resources.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation execute-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Wait until the `UPDATE` operation is complete, by using the the [wait stack-update-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-update-complete.html) AWS CLI command.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation wait stack-update-complete \
        --stack-name cfn-workshop-resource-importing
      :::

>     ```suggestion

Reviewer: ```suggestion
```

    --change-set-type IMPORT \
    --resources-to-import file://resources-import.txt \
    --template-body file://resource-importing.yaml \
    --parameters ParameterKey=Topic1Name,ParameterValue=Topic1
  :::
  1. Review the change set to make sure the correct resources are imported.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation describe-change-set \
    --stack-name cfn-workshop-resource-importing \
    --change-set-name cfn-workshop-resource-import-change-set
  :::
  1. Execute the change set to import the resources.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation execute-change-set \
    --stack-name cfn-workshop-resource-importing \
    --change-set-name cfn-workshop-resource-import-change-set
  :::
  1. Wait until the `IMPORT` operation is complete, by using the the [wait stack-import-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-import-complete.html) AWS CLI command.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation wait stack-import-complete \
    --stack-name cfn-workshop-resource-importing
  :::

> 

Author: We can try that however user might have to run a few times until they wait for this status.. The wait command pretty much does the similar job without having the need to run multiple times.

        --parameters ParameterKey=Topic1Name,ParameterValue=Topic1 ParameterKey=Topic2Name,ParameterValue=Topic2
      :::
      1. Review the change set to make sure the correct resources are imported.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation describe-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Execute the change set to import the resources.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation execute-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Wait until the `IMPORT` operation is complete, by using the the [wait stack-import-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-import-complete.html) AWS CLI command.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation wait stack-import-complete \
        --stack-name cfn-workshop-resource-importing
      :::

>     

Reviewer: Again not totally needed but informational - please mark conversation as resolved if you don't think it is needed

        --template-body file://resource-importing.yaml \
        --parameters ParameterKey=Topic2Name,ParameterValue=Topic2
      :::
      1. Review the change set to make sure the correct resources are removed.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation describe-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Execute the change set to import the resources.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation execute-change-set \
        --stack-name cfn-workshop-resource-importing \
        --change-set-name cfn-workshop-resource-import-change-set
      :::
      1. Wait until the `UPDATE` operation is complete, by using the the [wait stack-update-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-update-complete.html) AWS CLI command.
      :::code{language=shell showLineNumbers=false showCopyAction=true}
      aws cloudformation wait stack-update-complete \
        --stack-name cfn-workshop-resource-importing
      :::

>     

Reviewer: I think the second check to `1. Verify update complete by using the [describe-stack-resources]` is important as this is comparable to the Local Development steps and provides confirmation everything went okay

    --change-set-type IMPORT \
    --resources-to-import file://resources-import.txt \
    --template-body file://moving-resources.yaml \
    --parameters ParameterKey=Topic1Name,ParameterValue=Topic1
  :::
  1. Review the change set to make sure the correct resources are imported.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation describe-change-set \
    --stack-name cfn-workshop-moving-resources \
    --change-set-name cfn-workshop-moving-resources-change-set
  :::
  1. Execute the change set to import the resources.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation execute-change-set \
    --stack-name cfn-workshop-moving-resources \
    --change-set-name cfn-workshop-moving-resources-change-set
  :::
  1. Wait until the `IMPORT` operation is complete, by using the the [wait stack-import-complete](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/wait/stack-import-complete.html) AWS CLI command.
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  aws cloudformation wait stack-import-complete \
    --stack-name cfn-workshop-moving-resources
  :::

> 

Reviewer: Again `describe-stacks` may not be important but I think `1. Verify update complete by using the [describe-stack-resources]` is to verify

      aws cloudformation describe-stack-resources --stack-name cfn-workshop-moving-resources
      :::
      1. If the `describe-stack-resources` command was successfully sent, CloudFormation will return the stack resource information with `SNSTopic1`.
      :::code{language=json showLineNumbers=true showCopyAction=false}
{
  "StackResources": [
    {
      "StackName": "cfn-workshop-moving-resources",
      "StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-moving-resources/70f207d0-f459-11ed-9c39-123ae332d1a1",
      "LogicalResourceId": "SNSTopic1",
      "PhysicalResourceId": "arn:aws:sns:us-east-1:123456789012:Topic1.fifo",
      "ResourceType": "AWS::SNS::Topic",
      "Timestamp": "2023-05-17T02:21:15.205000+00:00",
      "ResourceStatus": "UPDATE_COMPLETE",
      "DriftInformation": {
        "StackResourceDriftStatus": "NOT_CHECKED"
      }
    }
  ]
}
  :::
  ::::
	::::tab{id="local" label="Local development"}  

1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
1. From **Create stack**, choose **With existing resources (import resources)**.

> 1. Read the **Import Overview** and choose **Next**.

Reviewer: ```suggestion
1. Read the **What you'll need** information and choose **Next**.
```

1. Make sure you are in the directory: `code/workspace/resource-importing`

> 2. Update the `resource-importing.yaml` template file to remove the `DeletionPolicy: Retain` line from the `SNSTopic2` resource definition, and save the template.

Reviewer: Don't we want to set `DeletionPolicy: Delete` so the resources are cleaned up?

Author: We are removing the Deletion Policy section in the template so that the resources are cleaned up upon deletion

Reviewer: If I remove `DeletionPolicy` I'm unable to update the stack - "no changes detected" and this is how I got around it, by explicitly setting `Delete`

Author: I have changed the command, earlier the errors were on the change sets, now that we are directly updating stacks it should be sufficient. I have tested this on the cloud 9 console

  :::::tabs{variant="container"}
::::tab{id="cloud9" label="Cloud9"}
    1. Update the **Stack** by using the following AWS CLI command
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation create-change-set \
          --stack-name cfn-workshop-resource-importing \
          --change-set-name cfn-workshop-resource-import-change-set \
          --change-set-type UPDATE \
          --template-body file://resource-importing.yaml
    :::
    1. Execute the change set to update the stack by using the following AWS CLI command
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation execute-change-set \
          --stack-name cfn-workshop-resource-importing \
          --change-set-name cfn-workshop-resource-import-change-set \

>     :::

Reviewer: I did the next step too fast and got an error, that's why I suggested this additional `wait stack-update-complete`

    aws cloudformation create-change-set \
        --stack-name cfn-workshop-resource-import-challenge \
        --change-set-name import-challenge --change-set-type IMPORT \
        --resources-to-import file://resources-import.txt \
        --template-body file://resource-import-challenge.yaml \
        --parameters ParameterKey=InstanceType,ParameterValue=t2.micro
    :::
    1. Execute the change set by using the following code
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation execute-change-set \
          --stack-name cfn-workshop-resource-import-challenge \
          --change-set-name import-challenge
    :::
    1. Wait unitl the `IMPORT` operation is complete by using the following AWS CLI command
    :::code{language=shell showLineNumbers=false showCopyAction=true}
    aws cloudformation wait stack-import-complete \
          --stack-name cfn-workshop-resource-import-challenge
    :::
      ::::
        ::::tab{id="LocalDevelopment" label="LocalDevelopment"}  
        1. Select the stack named `cfn-workshop-resource-import-challenge` and, from **Stack actions**, choose **Import resources into stack**.

>         1. Read the **what you'll need** information and choose **Next**.

Reviewer: ```suggestion
        1. Read the **What you'll need** information and choose **Next**.
```
